torch_xla_py/xla_dist.py

```diff
@@ -33,17 +33,55 @@ def __init__(self, client_workers, service_workers):
       client_workers: a list of ClientWorker objects.
       service_workers: a list of ServiceWorker objects.
     """
+    for client_worker in client_workers:
+      if not isinstance(client_worker, ClientWorker):
+        raise ValueError(
+            'client_workers argument must be a list of ClientWorker')
+    for service_worker in service_workers:
+      if not isinstance(service_worker, ServiceWorker):
+        raise ValueError(
+            'service_workers argument must be a list of ServiceWorker')
     self._client_workers = list(client_workers)
     self._service_workers = list(service_workers)
 
+
   def validate(self):
     """Validates the current cluster configuration.
 
     Raises:
       RuntimeError: If the cluster is misconfigured, this validation will
-        raise an error.
+        raise an error. For example, if the VMs are in different zones,
+        or not all of the CPU workers have the same size (number of CPU
+        cores, RAM size) we raise an exception. For TPUs we similarly
+        raise an exception if different zones or machine/accelerator_type.
     """
-    raise NotImplementedError()
+    if len(self._client_workers) == 0 or len(self._service_workers) == 0:
+      raise RuntimeError(
+        'Both client_workers and service_workers should not be empty')
+
+    if len(self._client_workers) != len(self._service_workers):
+      raise RuntimeError(
+          'The client_workers and service_workers must have a 1:1 mapping')
+
+    zones = {worker._zone for worker in self._client_workers}
+    zones.update({worker._zone for worker in self._service_workers})
```

Comment:
I think you don't need to make service_workers into a set before calling .update() since the update call will also remove duplicates. So this can be:
zones.update(worker._zone for worker in self._service_workers)

Reply:
So these 2 lines would become:

```
    zones = {worker._zone for worker in self._client_workers}
    zones.update(worker._zone for worker in self._service_workers)
```

Reply:
Done.